### PR TITLE
opt: fix bug where ON CONFLICT DO NOTHING ignores some input

### LINKED ISSUE #59125: opt: ON CONFLICT DO NOTHING ignores some input
Consider the following 
```
demo@127.0.0.1:26257/movr> CREATE TABLE uniq (
  k INT PRIMARY KEY,
  v INT UNIQUE,
  w INT UNIQUE,
  x INT,
  y INT DEFAULT 5,
  UNIQUE (x, y)
);
CREATE TABLE

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/movr> INSERT INTO uniq VALUES (1, 1, 1, 1, 1);
INSERT 1

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/movr> INSERT INTO uniq VALUES (1, 20, 20, 20, 20), (20, 1, 20, 20, 20), (20, 20, 20, 20, 20),
                        (20, 20, 1, 20, 20), (20, 20, 20, 1, 1) ON CONFLICT DO NOTHING;
INSERT 0

Time: 2ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/movr> SELECT * FROM uniq;
  k | v | w | x | y
----+---+---+---+----
  1 | 1 | 1 | 1 | 1
(1 row)

Time: 1ms total (execution 1ms / network 0ms)
```
There is one row in the second insert statement that did not have a conflict and therefore should have been inserted: `(20, 20, 20, 20, 20)`. However, it was not inserted. Postgres does successfully insert the row in this case, so this appears to be a bug.

### PR DESCRIPTION
**Pull request by rytaft**
Prior to this patch, it was possible for some valid input to an
`INSERT ... ON CONFLICT DO NOTHING` to be discarded. For example,
consider the following example:
```
CREATE TABLE uniq (
  k INT PRIMARY KEY,
  v INT UNIQUE,
  w INT UNIQUE,
  x INT,
  y INT DEFAULT 5,
  UNIQUE (x, y)
);

INSERT INTO uniq VALUES (1, 1, 1, 1, 1);

INSERT INTO uniq VALUES (1, 20, 20, 20, 20),
                        (20, 1, 20, 20, 20),
                        (20, 20, 20, 20, 20)
ON CONFLICT DO NOTHING;
```
Since row `(20, 20, 20, 20, 20)` does not conflict with the existing
row in uniq, it should be inserted. However, prior to this patch, all
three rows in the second `INSERT` statement were discarded.

This commit fixes the problem by applying the `upsert-distinct-on`
operators for each index after all conflicting rows are removed by
`anti-join`s.

Fixes #59125

Release note (bug fix): Fixed a bug in which some non-conflicting rows
provided as input to an `INSERT ... ON CONFLICT DO NOTHING` statement could
be discarded, and not inserted. This could happen in cases where the
table had one or more unique indexes in addition to the primary index,
and some of the rows in the input conflicted with existing values in one
or more unique index. This scenario could cause the rows that did not
conflict to be erroneously discarded. This has now been fixed.

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
Wow, nice find!

:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-:-MRQmE8w4_fo5wVdySWJ:b-nf031d)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/upsert, line 1258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQn2NI1ZARRX-_MW_P:-MRQn2NI1ZARRX-_MW_Q:b-18qg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/logictest/testdata/logic_test/upsert#L1258)):*
> ```
>   w INT UNIQUE,
>   x INT,
>   y INT DEFAULT 5,
> ```

[nit] the DEFAULT 5 is unnecessary
___
*[pkg/sql/logictest/testdata/logic_test/upsert, line 1266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQn4ba34weqYCUVc6l:-MRQn4ba34weqYCUVc6m:b17nxw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/logictest/testdata/logic_test/upsert#L1266)):*
> ```
> 
> statement ok
> INSERT INTO uniq VALUES (1, 20, 20, 20, 20), (20, 1, 20, 20, 20), (20, 20, 20, 20, 20),
> ```

[nit] It's hard to parse this example and associate values to columns. I would choose different ranges of values for each column, or better yet instead of integers use strings that incorporate the column name (`k1', 'x1', 'x2').

Might also be useful to add a comment with a note for each row (e.g. the second row must be discarded because of conflict on v).
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQoPNa36WW-ukUTOS5:-MRQoPNa36WW-ukUTOS6:buwi5t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/opt/optbuilder/testdata/upsert#L74)):*
> ```
>   w INT UNIQUE,
>   x INT,
>   y INT DEFAULT 5,
> ```

[nit] unnecessary DEFAULT 5

**Review by mgartner — APPROVED** [review]
:lgtm: Nice find! I'm glad the fix is so simple.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-:-MRQoZlH2S99WTht_De0:b-pd6hon)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball and @rytaft)

**Review by rytaft — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-:-MRRtEsp2dhR3kDUxiE6:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @mgartner)

___
*[pkg/sql/logictest/testdata/logic_test/upsert, line 1258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQn2NI1ZARRX-_MW_P:-MRRsg9zE7ATNBtYP1tE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/logictest/testdata/logic_test/upsert#L1258)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] the DEFAULT 5 is unnecessary
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/upsert, line 1266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQn4ba34weqYCUVc6l:-MRRsiQ5-Unhw6aE6NEx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/logictest/testdata/logic_test/upsert#L1266)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] It's hard to parse this example and associate values to columns. I would choose different ranges of values for each column, or better yet instead of integers use strings that incorporate the column name (`k1', 'x1', 'x2').

Might also be useful to add a comment with a note for each row (e.g. the second row must be discarded because of conflict on v).
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/upsert, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59147#-MRQoPNa36WW-ukUTOS5:-MRRt1hQAsEIndlHur1x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d66ea0942a35046b0a6e068cd4f0373fd83dee64/pkg/sql/opt/optbuilder/testdata/upsert#L74)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] unnecessary DEFAULT 5
</blockquote></details>

Done.
